### PR TITLE
Add Lagged Variable PostProcessor

### LINKED ISSUE #1522: [TASK] Add lagged parameter creation
--------
Issue Description
--------

**Is your feature request related to a problem? Please describe.**
There is no in RAVEN way to create lagged parameters in a HistorySet

**Describe the solution you'd like**
It would be useful if there was a PostProcessor that could take a variable in a HistorySet, and add some lagged parameters (like the previous hour, previous day etc).

**Describe alternatives you've considered**

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by joshua-cogliati-inl**
--------
Pull Request Description
--------
##### What issue does this change request address? 
Closes #1522 

##### What are the significant changes in functionality due to this change request?
This adds a new post processor.

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

File: doc/user_manual/postprocessor.tex [hunk 1]
```diff
@@ -1473,6 +1473,41 @@ \subsubsection{Interfaced}
 </Simulation>
 \end{lstlisting}
 
+\paragraph{Method: HistorySetDelay}
+
+This Post-Processor allows history sets to add delayed or lagged
+variables.  For example, if there a variable price that is set hourly,
+than new variable called price\_prev\_hour could be set by using a
+delay of -1.  This can be useful for training a ROM or other data
+analysis.
```
Comment (alfoa) [inline doc/user_manual/postprocessor.tex:1482]: Josh, please add an example here that precisely explains the Input DataObject and how the Output DataObject will get out. From the previous explanation I do not know how to use this PP and how to set up the output dataobject.
We need to be very verbose (with examples).

File: doc/user_manual/postprocessor.tex [hunk 2]
```diff
@@ -1473,6 +1473,41 @@ \subsubsection{Interfaced}
...
+
+In the \xmlNode{PostProcessor} input block, one or more of the following XML sub-nodes are required:
+
+\begin{itemize}
+\item \xmlNode{delay}, \xmlDesc{empty}, a delay node with the following required parameters:
+  \begin{itemize}
+  \item \xmlAttr{original}, the variable to start with
```
Comment (alfoa) [inline doc/user_manual/postprocessor.tex:1489]: This is not the format of the manual.
Example:

 ```\item \xmlAttr{VAR}, \xmlDesc{required string attribute}, user-defined VARIABLE name etc etc```

File: doc/user_manual/postprocessor.tex (hunk 1 above)
Comment (alfoa) [reply doc/user_manual/postprocessor.tex:1482]: Maybe in conjunction with the example you have below.

File: doc/user_manual/postprocessor.tex [hunk 3]
```diff
@@ -1473,6 +1473,41 @@ \subsubsection{Interfaced}
...
+  \item \xmlAttr{new}, the new variable to create
+  \item \xmlAttr{steps}, the delay (if negative) or steps into the future (if positive) to use for the new variable (so -1 gives the previous, 1 gives the next)
+  \item \xmlAttr{default}, the value to use for cases where there is no previous or next value.
```
Comment (alfoa) [inline doc/user_manual/postprocessor.tex:1492]: please be verbose here... "there is no previous or next value". What do you mean? in terms of perfect match using the step? need clear explanation for the user

File: framework/PostProcessorFunctions/HistorySetDelay.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
+
+"""
+This is to implement a delay or lagged parameters in a HistorySet
```
Comment (alfoa) [inline framework/PostProcessorFunctions/HistorySetDelay.py:3]: standard of docstring

File: framework/PostProcessorFunctions/HistorySetDelay.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+import copy
+import numpy as np
+
+from PostProcessorInterfaceBaseClass import PostProcessorInterfaceBase, CheckInterfacePP
+from utils import InputData, InputTypes
+
+class HistorySetDelay(PostProcessorInterfaceBase):
+  """
+  Class to get lagged or delayed data out of a history set.
```
Comment (alfoa) [inline framework/PostProcessorFunctions/HistorySetDelay.py:14]: standard

File: framework/PostProcessorFunctions/HistorySetDelay.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+      Method to get a reference to a class that specifies the input data for
+      class cls.
+      @ In, cls, the class for which we are retrieving the specification
+      @ Out, inputSpecification, InputData.ParameterInput, class to use for
+        specifying input of cls.
+    """
+    inputSpecification = super().getInputSpecification()
+    inputSpecification.setCheckClass(CheckInterfacePP("HistorySetDelay"))
+    delayClass = InputData.parameterInputFactory("delay", InputTypes.StringType,
+                                                 descr="Adds a delay variable")
```
Comment (alfoa) [inline framework/PostProcessorFunctions/HistorySetDelay.py:29]: verbosity of the description

Comment (alfoa) [reply framework/PostProcessorFunctions/HistorySetDelay.py:29]: those descriptions will be used for future automatic documentation. they must be of "manual grade"

File: framework/PostProcessorFunctions/HistorySetDelay.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+                        descr="Steps to offset (-1 is previous step)")
+    delayClass.addParam("default", InputTypes.FloatType, True,
+                        descr="Default value to use for unavailable steps")
+    inputSpecification.addSub(delayClass, InputData.Quantity.one_to_infinity)
+    inputSpecification.addSub(InputData.parameterInputFactory("method", contentType=InputTypes.StringType))
+    return inputSpecification
+
+  def initialize(self):
+    """
+      Method to initialize the HistorySetDelay
```
Comment (alfoa) [inline framework/PostProcessorFunctions/HistorySetDelay.py:44]: standard docstring

File: tests/framework/PostProcessors/InterfacedPostProcessor/test_historySetDelay.xml [hunk 8]
```diff
@@ -0,0 +1,70 @@
+<?xml version="1.0" ?>
+<Simulation verbosity="debug">
+  <TestInfo>
+    <name>framework/PostProcessors/InterfacedPostProcessor/HistorySetDelay</name>
+    <author>cogljj</author>
+    <created>2021-04-19</created>
+    <classesTested>InterfacedPostProcessor</classesTested>
+    <description>
+      Tests of the HistorySetDelay interfaced post-processor
```
Comment (alfoa) [inline tests/framework/PostProcessors/InterfacedPostProcessor/test_historySetDelay.xml:9]: Can you improve the description answering to the following questions:
what do we test? What the expected outcomes? input and outputs? the model used?

**Review by alfoa — CHANGES_REQUESTED** [review]
The interfacePostprocessor here should be converted to be a normal PostProcessor (please talk to @wangcj05 about the new structure)

**Comment by joshua-cogliati-inl** [comment]
> The interfacePostprocessor here should be converted to be a normal PostProcessor (please talk to @wangcj05 about the new structure)

Hm, HistorySetSync doesn't seem to be converted in #1512  Where is an example of a converted InterfacePostProcessor I can look at?

**Comment by wangcj05** [comment]
> > The interfacePostprocessor here should be converted to be a normal PostProcessor (please talk to @wangcj05 about the new structure)
> 
> Hm, HistorySetSync doesn't seem to be converted in #1512 Where is an example of a converted InterfacePostProcessor I can look at?

I have converted the "RiskMeasuresDiscrete" in PR: https://github.com/idaholab/raven/pull/1512/files

**Comment by joshua-cogliati-inl** [comment]
> > > The interfacePostprocessor here should be converted to be a normal PostProcessor (please talk to @wangcj05 about the new structure)
> > 
> > 
> > Hm, HistorySetSync doesn't seem to be converted in #1512 Where is an example of a converted InterfacePostProcessor I can look at?
> 
> I have converted the "RiskMeasuresDiscrete" in PR: https://github.com/idaholab/raven/pull/1512/files

The method run doesn't seem to be modified:

https://github.com/idaholab/raven/pull/1512/files#diff-1d3a8ec5e63f5ed917bae8cea26d940241b149fcfda0e001b29ff32d871c83e7

**Comment by wangcj05** [comment]
> > > > The interfacePostprocessor here should be converted to be a normal PostProcessor (please talk to @wangcj05 about the new structure)
> > > 
> > > 
> > > Hm, HistorySetSync doesn't seem to be converted in #1512 Where is an example of a converted InterfacePostProcessor I can look at?
> > 
> > 
> > I have converted the "RiskMeasuresDiscrete" in PR: https://github.com/idaholab/raven/pull/1512/files
> 
> The method run doesn't seem to be modified:
> 
> https://github.com/idaholab/raven/pull/1512/files#diff-1d3a8ec5e63f5ed917bae8cea26d940241b149fcfda0e001b29ff32d871c83e7

The base class will be changed. Based on the new structure, the interfacePP class, the PPInterfaces, the PPInterfaceBaseClass are not needed anymore. There is only minor changes in the methods by using inheritance.

**Comment by joshua-cogliati-inl** [comment]
Well, I attempted to convert it to a post processor in: https://github.com/idaholab/raven/commit/f4244c68a9c590881420f601029b66ed3eb0e244

File: tests/framework/PostProcessors/InterfacedPostProcessor/test_historySetDelay.xml (hunk 8 above)
Comment (joshua-cogliati-inl) [reply tests/framework/PostProcessors/InterfacedPostProcessor/test_historySetDelay.xml:9]: Fixed.

File: framework/PostProcessorFunctions/HistorySetDelay.py (hunk 7 above)
Comment (joshua-cogliati-inl) [reply framework/PostProcessorFunctions/HistorySetDelay.py]: Fixed.

File: doc/user_manual/postprocessor.tex (hunk 2 above)
Comment (joshua-cogliati-inl) [reply doc/user_manual/postprocessor.tex:1489]: Fixed.

File: doc/user_manual/postprocessor.tex (hunk 1 above)
Comment (joshua-cogliati-inl) [reply doc/user_manual/postprocessor.tex:1482]: Added a bit more here, and expanded the example to include the step and the data objects.

**Comment by joshua-cogliati-inl** [comment]
This should be ready to review again.

File: tests/framework/PostProcessors/HistorySetDelay/test_historySetDelay.xml [hunk 9]
```diff
@@ -0,0 +1,72 @@
+<?xml version="1.0" ?>
+<Simulation verbosity="debug">
+  <TestInfo>
+    <name>framework/Models/PostProcessors/HistorySetDelay</name>
+    <author>cogljj</author>
+    <created>2021-04-19</created>
+    <classesTested>InterfacedPostProcessor</classesTested>
```
Comment (wangcj05) [inline tests/framework/PostProcessors/HistorySetDelay/test_historySetDelay.xml:7]: change InterfacedPostProcessor to HistorySetDelay.

File: tests/framework/PostProcessors/HistorySetDelay/test_historySetDelay.xml [hunk 10]
```diff
@@ -0,0 +1,72 @@
+<?xml version="1.0" ?>
+<Simulation verbosity="debug">
+  <TestInfo>
+    <name>framework/Models/PostProcessors/HistorySetDelay</name>
```
Comment (wangcj05) [inline tests/framework/PostProcessors/HistorySetDelay/test_historySetDelay.xml:4]: This should be the test name, i.e. framework/PostProcessors.HistorySetDelay

**Review by wangcj05 — CHANGES_REQUESTED** [review]
Some minor comments.

File: framework/Models/PostProcessors/HistorySetDelay.py [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+    """
+      Method to get a reference to a class that specifies the input data for
+      class cls.
+      @ In, cls, the class for which we are retrieving the specification
+      @ Out, inputSpecification, InputData.ParameterInput, class to use for
+        specifying input of cls.
+    """
+    inputSpecification = super().getInputSpecification()
+    #inputSpecification.setCheckClass(CheckInterfacePP("HistorySetDelay"))
+    delayClass = InputData.parameterInputFactory("delay", InputTypes.StringType,
```
[the remark this answers is not in the document]
Comment (joshua-cogliati-inl) [reply framework/Models/PostProcessors/HistorySetDelay.py:46]: Thanks.

**Review by wangcj05 — APPROVED** [review]
changes are good.

**Review by alfoa — APPROVED** [review]
Comments addressed

**Comment by alfoa** [comment]
Checklist passed...

Comments addressed..

Approved to merge.

**Comment by alfoa** [comment]
Merging...

**Pull request merged**